### PR TITLE
EZP-25579: SystemInfo view

### PR DESCRIPTION
**Pull request by bdunogier**
> https://jira.ez.no/browse/EZP-25579
> Working prototype, with all the hard requirements
> PlatformUI integration: https://github.com/ezsystems/PlatformUIBundle/pull/532
> ezplatform integration: https://github.com/ezsystems/ezplatform/compare/ezp25579-view_api_integration

### Contents
- `SystemInfoViewController` + `viewInfoAction`
- `/ez-support-tools/view/info/{identifier}` route. Identifier is the name of the class minus SystemInfo, underscorized: `PhpSystemInfo` :arrow_right: `php`
- `SystemInfoView` object + builder
- `SystemInfo\Identifier` view matcher

### Status
`/ez-support-tools/view/php` matches the SystemInfoController view action.
The template (property of the view) used to render the SystemInfoView is set depending on the view matchers configuration.

#### View configuration
The template used by each info type can be configured as for  `content_view`:

```
parameters:
    ezsettings.default.system_info_view:
        full:
            php:
                template: "ez-support-tools/info/php.html.twig"
                match:
                    SystemInfo\Identifier: "php"
            hardware:
                template: "ez-support-tools/info/hardware.html.twig"
                match:
                    SystemInfo\Identifier: "hardware"
```

#### View templates
The viewed `SystemInfo` object is provided as the `info` variable.

We don't intend to provide system info outside of a PlatformUI context for the time being. Therefore, the *pjax* view templates should be added to PlatformUI.

#### Permissions & security

How was this handled in PlatformUI ?
First, I think we can _maybe_ postpone that part: the package is still a 0.x, and we can easily document that in the readme.

As for implementing this, I'd lean towards real policies: we can have services around collectors that do check permissions using custom policies.

### Tasks
- [x] PlatformUI templates. The idea is that PlatformUI would use its own templates to implement pjax
- [x] Tests
- [x] Semantic config for system_info_view. But is it a hard requirement ?
- [x] Think about security/permissions

### CODE REVIEW

File: Resources/config/services.yml [hunk 1]
```diff
@@ -16,3 +16,10 @@ services:
     support_tools.info_collector.system_info.ezc_system_info:
         class: ezcSystemInfo
         factory: [EzSystems\EzSupportToolsBundle\SystemInfo\EzcSystemInfoFactory, buildEzcSystemInfo]
+
+    support_tools.view.system_info_view_builder:
+        class: EzSystems\EzSupportToolsBundle\View\SystemInfoViewBuilder
+        arguments: [@support_tools.info_collectors.ezc_hardware]
```
Comment (bdunogier) [inline Resources/config/services.yml]: Note to self: probably not final.

File: View/SystemInfoView.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\EzSupportToolsBundle\View;
+
+use eZ\Publish\Core\MVC\Symfony\View\BaseView;
+use eZ\Publish\Core\MVC\Symfony\View\View;
+
+class SystemInfoView extends BaseView implements View
```
Comment (bdunogier) [inline View/SystemInfoView.php]: Note to self: phpdoc.

File: View/SystemInfoViewBuilder.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return $argument === 'support_tools.view.controller:viewInfoAction';
+    }
+
+    public function buildView(array $parameters)
+    {
+        $view = new SystemInfoView();
+        $view->setInfo($this->infoCollector->build());
+        $view->setTemplateIdentifier(
+            $this->toTemplateIdentifier($view->getInfo())
```
Comment (bdunogier) [inline View/SystemInfoViewBuilder.php]: Note to self: must be changed.

The template identifier should come from the ViewMatchers (to @glye : executed by the Configurator, in the View Request Listener).

It could be interesting to base the default template name on a convention, probably from the view builder (computed based on the class name, as it is now), with a convention based directory (`HardwareSystemInfo` => `ez-support-tools/hardware_system_info.html.twig`).

We are kind of getting close from the concept we had in legacy, where the _requested template_ was part of the view rule. We don't have this feature yet (matchers can't/don't test that).

Shorter version: if a default template is set by convention, based on the rendered info, in the view object, how could we take advantage of that in view rules ?

File: View/SystemInfoViewBuilder.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\EzSupportToolsBundle\View;
+
+use Doctrine\Common\Inflector\Inflector;
+use eZ\Publish\Core\MVC\Symfony\View\Builder\ViewBuilder;
+use EzSystems\EzSupportToolsBundle\SystemInfo\Collector\SystemInfoCollector;
+
+class SystemInfoViewBuilder implements ViewBuilder
```
Comment (bdunogier) [inline View/SystemInfoViewBuilder.php]: Should probably also implement `CachableView`.

Comment (glye) [inline View/SystemInfoViewBuilder.php]: Agree, but.. how should this cache be expired?

File: View/SystemInfoViewBuilder.php (hunk 3 above)
Comment (glye) [inline View/SystemInfoViewBuilder.php]: > if a default template is set by convention, based on the rendered info, in the view object, how could we take advantage of that in view rules?

Isn't that answered by `toTemplateIdentifier()`? It's somewhat foggy to me, though I'm all for convention rather than configuration.

File: View/SystemInfoViewBuilder.php (hunk 4 above)
Comment (bdunogier) [inline View/SystemInfoViewBuilder.php]: Do we need to expire it, as long as there is a TTL ?

File: View/SystemInfoViewBuilder.php (hunk 3 above)
Comment (bdunogier) [inline View/SystemInfoViewBuilder.php]: Well, so far we _only_ did configuration.

The `toTemplateIdentifier()` indeed does that. On the other hand, it's hack, and it's not done like that at all in the places where we use this API.

File: Controller/SystemInfoController.php [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?php
+/**
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace EzSystems\EzSupportToolsBundle\Controller;
+
+use EzSystems\EzSupportToolsBundle\View\SystemInfoView;
+
+/**
+ * Do we need that controller at all ?
```
Comment (bdunogier) [inline Controller/SystemInfoController.php]: Needs answer before merge.

**Comment by bdunogier** [comment]
The critical path should be working now. Updating the description with current state and how to use.

**Comment by bdunogier** [comment]
Added some stuff about permissions & security, with questions.

File: Controller/SystemInfoController.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+ */
+class SystemInfoController
+{
+    public function viewInfoAction(SystemInfoView $view)
```
Comment (glye) [inline Controller/SystemInfoController.php]: Isn't this parameter a string - `systemInfoIdentifier` as defined in routing.yml? Or is some automagic taking care of it?

**Comment by glye** [comment]
> Create the matching template in app/Resources/view. Example: app/Resources/ez-support-tools/info/hardware.html.twig.

`app/Resources/view/ez-support-tools/info/hardware.html.twig` then..?
(Alternatively they can be put in the bundle, right?)

**Comment by bdunogier** [comment]
> (Alternatively they can be put in the bundle, right?)

Yes, indeed. But I don't think we should right now, because of scope.
Our scope is to provide this feature withinPlatformUI / pjax. We don't plan on providing SystemInfo outside of this context.

I'll push very soon override templates _in PlatformUI_, it will make more sense :-)

File: Controller/SystemInfoController.php (hunk 6 above)
Comment (bdunogier) [inline Controller/SystemInfoController.php]: Automagic it is: the "View API".

**Comment by glye** [comment]
> How was this handled in PlatformUI?

It just requires the `setup/system_info` policy, it seems.

**Comment by bdunogier** [comment]
> Did someone edit my comment? I don't recall the above line :)
Actually I did. Bad reflex, might have been (too) late. Hit 'edit' instead of 'reply'. The buttons are sooo close from each other :D

**Comment by bdunogier** [comment]
About permissions, we should aim, as soon as possible, to having a service layer that includes permissions (could be a decorator).

This principle only exists in the Public API, but there are no real reasons why we'd do things differently in an extension. Now the question is what does this service _serve_ exactly ? The SystemInfo view isn't our end goal. It's only meant to be used to render the SystemInfo objects from a "Report". Should we build that service around reports, and use that in our View ?

**Comment by glye** [comment]
Re: permissions, what granularity do we need? Per infocollector, per report, or a single one for the whole thing (probably not).

**Comment by bdunogier** [comment]
> Re: permissions, what granularity do we need?

We can be agile about it. For now, we should just stick to what we used before.
Any opinion on the service layer aspect ?

**Comment by glye** [comment]
I thought we'd need some idea about intended granularity before designing something that deals with permissions. Anyway, what else should the service layer do?

**Comment by bdunogier** [comment]
> Anyway, what else should the service layer do?

It should do what we need it to do ? :-)

But before we dig into that, I suggest that we merge this PR without handling it, and create a follow-up before we merge.

**Comment by bdunogier** [comment]
Created follow-ups for [permissions](https://jira.ez.no/browse/EZP-25614) and [implementation of PJAX templates in PlatformUI](https://jira.ez.no/browse/EZP-25615).
